Mathlib/RingTheory/DedekindDomain/Factorization.lean

```diff
@@ -177,3 +223,289 @@ theorem finprod_heightOneSpectrum_factorization_coe (I : Ideal R) (hI : I ≠ 0)
+
+/-- For every `n ∈ ℕ` and every ideal `I`, `val_v(I^n) = n*val_v(I)`. -/
+theorem count_pow (n : ℕ) (I : FractionalIdeal (nonZeroDivisors R) K) :
+    count K v (I ^ n) = n * count K v I := by
+  induction' n with n h
+  · rw [pow_zero, ofNat_zero, MulZeroClass.zero_mul, count_one]
+  · rw [pow_succ, count_mul']
+    by_cases hI : I = 0
+    · have h_neg : ¬(I ≠ 0 ∧ I ^ n ≠ 0) := by
+        rw [not_and, Classical.not_not, Ne.def]
+        intro h
+        exact absurd hI h
+      rw [if_neg h_neg, hI, count_zero, MulZeroClass.mul_zero]
+    · rw [if_pos (And.intro hI (pow_ne_zero n hI)), h, Nat.succ_eq_add_one, Nat.cast_add,
+        Nat.cast_one]
+      ring
+
+/-- `val_v(v) = 1`, when `v` is regarded as a fractional ideal. -/
+theorem count_self : count K v (v.asIdeal : FractionalIdeal (nonZeroDivisors R) K) = 1 := by
+  have hv : (v.asIdeal : FractionalIdeal (nonZeroDivisors R) K) ≠ 0 := coeIdeal_ne_zero.mpr v.ne_bot
+  have h_self : (v.asIdeal : FractionalIdeal (nonZeroDivisors R) K) =
+      spanSingleton (nonZeroDivisors R) ((algebraMap R K) 1)⁻¹ * ↑v.asIdeal := by
+    rw [(algebraMap R K).map_one, inv_one, spanSingleton_one, one_mul]
+  have hv_irred : Irreducible (Associates.mk v.asIdeal) := by apply v.associates_irreducible
+  rw [count_well_defined K v hv h_self, Associates.count_self hv_irred, Ideal.span_singleton_one,
+    ← Ideal.one_eq_top, Associates.mk_one, Associates.factors_one, Associates.count_zero hv_irred,
+    ofNat_zero, sub_zero, ofNat_one]
+
+/-- `val_v(v) = n` for every `n ∈ ℕ`. -/
+theorem count_pow_self (n : ℕ) :
+    count K v ((v.asIdeal : FractionalIdeal (nonZeroDivisors R) K) ^ n) = n := by
+  rw [count_pow, count_self, mul_one]
+
+/-- `val_v(I⁻ⁿ) = -val_v(Iⁿ)` for every `n ∈ ℤ`. -/
+theorem count_inv (n : ℤ) (I : FractionalIdeal (nonZeroDivisors R) K) :
+    count K v (I ^ (-n)) = - count K v (I ^ n) := by
+  by_cases hI : I = 0
+  · by_cases hn : n = 0
+    · rw [hn, neg_zero, zpow_zero, count_one, neg_zero]
+    · rw [hI, zero_zpow n hn, zero_zpow (-n) (neg_ne_zero.mpr hn), count_zero, neg_zero]
+  · rw [eq_neg_iff_add_eq_zero, ← count_mul K v (zpow_ne_zero _ hI) (zpow_ne_zero _ hI),
+      ← zpow_add₀ hI, neg_add_self, zpow_zero]
+    exact count_one K v
+
+/-- `val_v(Iⁿ) = n*val_v(I)` for every `n ∈ ℤ`. -/
+theorem count_zpow (n : ℤ) (I : FractionalIdeal (nonZeroDivisors R) K) :
+    count K v (I ^ n) = n * count K v I := by
+  cases' n with n
+  · rw [ofNat_eq_coe, zpow_ofNat]
+    exact count_pow K v n I
+  · rw [negSucc_coe, count_inv, zpow_ofNat, count_pow]
+    ring
+
+/-- `val_v(v) = n` for every `n ∈ ℤ`. -/
+theorem count_zpow_self (n : ℤ) :
+    count K v ((v.asIdeal : FractionalIdeal (nonZeroDivisors R) K) ^ n) = n := by
+  rw [count_zpow, count_self, mul_one]
+
+/-- If `v ≠ w` are two maximal ideals of `R`, then `val_v(w) = 0`. -/
+theorem count_maximal_coprime (w : HeightOneSpectrum R) (hw : w ≠ v) :
```

Comment:
Should there be a lemma saying `count K v (w.asIdeal : _) = if w = v then 1 else 0`?

Reply:
Added as `count_maximal`.